File: sonar-plugin-api/src/main/java/org/sonar/api/qualitymodel/CharacteristicProperty.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2012 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Sonar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package org.sonar.api.qualitymodel;

import javax.persistence.*;

/**
 * @since 2.3
 */
@Entity
@Table(name = "characteristic_properties")
public final class CharacteristicProperty {

  @Id
  @Column(name = "id")
  @GeneratedValue
  private Integer id;

  @Column(name = "kee", nullable = true, length = 100)
  private String key;

  @Column(name = "value", nullable = true)
  private Double value;

  @Column(name = "text_value", nullable = true, length = 4000)
  private String textValue;

  @ManyToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "characteristic_id", updatable = true, nullable = false)
  private Characteristic characteristic;

  /**
   * Use the factory method create()
   */
  CharacteristicProperty() {
  }

  public static CharacteristicProperty create(String key) {
    return new CharacteristicProperty().setKey(key);
  }

  public Integer getId() {
    return id;
  }

  CharacteristicProperty setId(Integer i) {
    this.id = i;
    return this;
  }

  public String getKey() {
    return key;
  }

  public CharacteristicProperty setKey(String s) {
    this.key = s;
    return this;
  }

  public String getTextValue() {
    return textValue;
  }

  public Double getValue() {
    return value;
  }

  public Long getValueAsLong() {
    if (value!=null) {
      return value.longValue();
    }
    return null;
  }

  public CharacteristicProperty setTextValue(String s) {
    this.textValue = s;
    return this;
  }

  public CharacteristicProperty setValue(Double d) {
    this.value = d;
    return this;
  }

  Characteristic getCharacteristic() {
    return characteristic;
  }

  CharacteristicProperty setCharacteristic(Characteristic c) {
    this.characteristic = c;
    return this;
  }
}

File: sonar-server/src/main/webapp/WEB-INF/app/models/characteristic_property.rb
#
# Sonar, entreprise quality control tool.
# Copyright (C) 2008-2013 SonarSource
# mailto:contact AT sonarsource DOT com
#
# SonarQube is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# SonarQube is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
class CharacteristicProperty < ActiveRecord::Base
  KEY_MAX_SIZE=100

  PROPERTY_REMEDIATION_FUNCTION = "remediationFunction";
  PROPERTY_REMEDIATION_FACTOR = "remediationFactor";
  PROPERTY_OFFSET = "offset";

  FUNCTION_CONSTANT = "constant_resource";
  FUNCTION_LINEAR = "linear";
  FUNCTION_LINEAR_WITH_OFFSET = "linear_offset";
  FUNCTION_LINEAR_WITH_THRESHOLD = "linear_threshold";

  DAY = "d"
  HOUR = "h"
  MINUTE = "mn"

  belongs_to :characteristic
  validates_length_of :kee, :in => 1..KEY_MAX_SIZE, :allow_blank => false
  
  def key
    kee
  end

  def constant?
    text_value == FUNCTION_CONSTANT
  end

  def linear?
    text_value == FUNCTION_LINEAR
  end

  def linearWithThreshold?
    text_value == FUNCTION_LINEAR_WITH_THRESHOLD
  end

  def linearWithOffset?
    text_value == FUNCTION_LINEAR_WITH_OFFSET
  end

end
